### ISSUE TITLE
Promregator running in PCF - Issues with /discovery - promregator.discovery.port

### ISSUE THREAD

**Issue by WilliamLHarms** [promregator/promregator#189]
## Environment
Implementation-Title: Promregator
Implementation-Version: 0.8.0

Running promregator within PCF PAS 2.8

From the prometheus node, doing a simple wget on a cron to build a discovery file.  
## Summary / Problem Statement
The discovery endpoint has a port, 8080 in the config.  When trying to curl to the https://route:8080/singleTargetMetrics/guids/instance, I get a cert error:
curl: (35) SSL received a record that exceeded the maximum permissible length.

When I try to browse there I receive:  ERR_SSL_PROTOCOL_ERROR

When I try to browse to the discovery route without the port, everything works fine.  

I am also reaching out to our PCF Platform team to see if something can be done on their side to help open the port.

Documents I consulted: 
https://github.com/promregator/promregator/blob/master/docs/config.md#option-promregatordiscoveryport-optional

Code I reviewed:
https://github.com/promregator/promregator/blob/bf0e2ac0b33065f474e257a872f8f03202f8defa/src/main/java/org/cloudfoundry/promregator/endpoint/DiscoveryEndpoint.java#L141

^^^ doesn't look like we can turn that off...   Thoughts on adding an additional property that will enable the default port to be off in the targets string?  My other option is to write a post process from the wget that will yank the port out of the file.

## Observed Behavior
Running the following cf push manifest:

```yaml
---
applications:
- name: promregator
  path: promregator.jar
  memory: 1g
  instances: 1
  env:
    JAVA_OPTS: '-Dspring.config.location=file:$PWD/promregator.yml'
    SKIPVERIFY: true
    spring.application.name: promregator
```

With this promregator yaml config:
```yaml
server:
  port: 50082
cf:
  api_host: api.sys.ap01.pcf.company.com
  username: SECRET
  password: SECRET
  cache.type: CAFFEINE
  cache.timeout.application: 600
  connectionPool:
    size: 32
  request.timeout.appSummary: 6000
  request.timeout.appInSpace: 3000
logging:
  level:
    org:
      cloudfoundry:
        promregator: ERROR
promregator:
  scraping:
    threads: 14
  authenticator:
    type: none
  discovery:
    hostname: promregator.apps.ap01.pcf.company.com
  targets:
```

Tomcat looks good:  o.s.b.w.embedded.tomcat.TomcatWebServer : Tomcat started on port(s): 8080 (http) with context path ''

Yields this discovery json.. Note the port... 
```json
{
"targets": [
"promregator.apps.ap01.pcf.company.com:8080"
],
"labels": {
"__meta_promregator_target_path": "/singleTargetMetrics/xxxxx-yyyy-zzzz-9999-888888888/0",
"__meta_promregator_target_orgName": "some-org",
"__meta_promregator_target_spaceName": "some-space",
"__meta_promregator_target_applicationName": "some-app",
"__meta_promregator_target_applicationId": "xxxxx-yyyy-zzzz-9999-888888888",
"__meta_promregator_target_instanceNumber": "0",
"__meta_promregator_target_instanceId": "xxxxx-yyyy-zzzz-9999-888888888:0",
"__metrics_path__": "/singleTargetMetrics/xxxxx-yyyy-zzzz-9999-888888888/0"
}
}
```

When I curl from the prometheus box to the singleTargetMetrics route, nothing works.  When I take the port off I'm good.

## Expected Behavior
The ability to remove a port reference from the promregator on discovery.

fyi - @eaglerainbow

**Comment by eaglerainbow** [comment]
Hi @WilliamLHarms, 

I think I got the problem! Good analysis!
Could you do me a favor and try setting 
``` 
promregator.discovery.port=443
```
? That might already do the trick, as the SSL's default port is 443.

That would work, perhaps adding another comment in [the documentation](https://github.com/promregator/promregator/blob/master/docs/config.md#option-promregatordiscoveryport-optional) could have prevented that from happening. If it does not work, we should understand why this is the case...

**Comment by eaglerainbow** [comment]
erhm... Most likely this also wouldn't work: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#file_sd_config states no way to change the scheme of what you scrape: That is to say, you can't switch to HTTPS-based scraping using `file_sd_config`. I think we stumbled here over a Prometheus' feature request... 

What happens if you say `promregator.discovery.port=80`. Does the webdispatcher's redirect (HTTP Status code 302) work here as well or is Prometheus' client to weak for that?

**Comment by eaglerainbow** [comment]
BTW: Be careful, if you expose the discovery endpoint to the public... You should then protect it by some authentication scheme (cf. https://github.com/promregator/promregator/blob/af3893bb2090fe10089ed1d14ddd43ebe62dabf0/docs/config.md#subgroup-promregatorauthentication)

**Comment by WilliamLHarms** [comment]
@eaglerainbow  - sure I'll give promregator.discovery.port=443 a whirl.. But after chatting with our PCF Platform Team admins.. Any port references won't work.  It has to do with how they set up the diego cells.

We've kept the entire config the same as above, but to get us working in prod, I added a little sed at the end to yoink the :8080 out of the targets.

```
*/2 * * * * wget https://promregator-org_name.apps.ap01.pcf.company.com/discovery -O /etc/prometheus/discovery/foundation_discovery.json;sed -i 's/:8080//g' /etc/prometheus/discovery/foundation_discovery.json
 ```
Yea I'll also check out those docs on adding some auth to discovery.  Might want to do that :)  Right now, we're protected by firewalls.  The space we're deploying to isn't accessable from the public.

**Comment by eaglerainbow** [comment]
> I added a little sed at the end to yoink the :8080 out of the targets.

And that works? If that works, then setting port to 443 should do the trick as well...

**Comment by WilliamLHarms** [comment]
Yep, we've been running the discovery with sed for about a week now.  Will get to the test with the 443 port sometime next week.  We're in the middle of a very busy week, and stability is key right now.

**Comment by eaglerainbow** [comment]
No rush. I do understand that Black Friday Week is much more important than this here!

**Comment by WilliamLHarms** [comment]
apologies for the late reply.. I didn't have much luck on 443 :(

**Comment by eaglerainbow** [comment]
Now I have to apology for late response: too many things in parallel last where here... And the next week also doesn't look bright.
Re-reading the issue to refresh my memory, it triggers some more questions. You *might* be running into the issue we are currently also battling with #195 . Therefore, let me briefly check: What's the route that you are referring to? Is the route's domain flagged as "internal"? If yes, I don't think your `sed`-approach is what you want (you'd be running into https://github.com/promregator/promregator/pull/195#issuecomment-787519316 then).

What also confuses me now heavily is that you have written
``` 
Tomcat looks good: o.s.b.w.embedded.tomcat.TomcatWebServer : Tomcat started on port(s): 8080 (http) with context path ''
```
which indicates that Promregator would be running on 8080, although your config file states
```
server:
  port: 50082
```
so Tomcat should be running on 50082.

Could you cross-check?

PS: I currently have a little trouble with knotting everything together properly (i.e. where is what, which port is used where, ...). Perhaps it makes sense - if you are using non-internal domains - to also go through the addresses and expected ports together. We might be missing there something...

**Comment by WilliamLHarms** [comment]
Thanks @eaglerainbow  - private message me if you'd like to do a one on one conversation and perhaps share a screen or two over teams.  I'll meet at a time convenient for you.  Also, I'm sure a few of my teammates would like to meet you.

It the meantime I'll dig into the check you asked for above.

**Comment by WilliamLHarms** [comment]
To clarify configs.. 

CF Manifest
`
---
applications:
- name: promregator-foundation-team
  build-pack: java_buildpack
  path: promregator-foundation-team.jar
  memory: 1g
  instances: 1
  env:
    JAVA_OPTS: '-Dspring.config.location=file:$PWD/app_configs/promregator-foundaiton-team.yml'
    PUBLISH_PORT: 443
    SKIPVERIFY: true
    spring.application.name: promregator-foundation-team
`

Port config on promregator.yml

`
server:
  port: 443
cf:
  api_host: our.cf.api.host
  username: usr
  password: pswd
  cache.type: CAFFEINE
  cache.timeout.application: 600
  connectionPool:
    size: 32
  request.timeout.appSummary: 6000
  request.timeout.appInSpace: 3000
logging:
  level:
    org:
      cloudfoundry:
        promregator: ERROR
promregator:
  scraping:
    threads: 14
  authenticator:
    type: none
  discovery:
    hostname: promregator-foundtaion-team.pcf.company.com
  metrics:
    internal: true
    `
    
    promregator/discovery
    `"targets": [
"promregator-foundation-team.pcf.company.com:8080"
],`

**Comment by eaglerainbow** [comment]
Great if that works!
I now try to understand - and why it works:
* I am not aware that the option `PUBLISH_PORT: 443` would do anything. Spring-boot does not react on this environment variable to my knowledge. So, this is a nil operation. I would expect that you still read something like `Tomcat started on port(s): 8080 (http) with context path ''` in your logs.
* The `server.port: 443` option in promregator.yaml would make your promregator instance listen on port container's port 443 (see also https://www.baeldung.com/spring-boot-change-port ). That might be okay, but could also fail (depending on the diego's configuration whether the container's vcap service user is allowed to run on privileged ports or not - apparently, at your's it may).
* Setting `promregator.discovery.hostname: promregator-foundtaion-team.pcf.company.com` would enforce Promregator to - whatever still it may think - write `promregator-foundtaion-team.pcf.company.com` (typo?) as host names into the targets of the discovery descriptor.
* The discovery endpoint tells then you may scrape at endpoints at `promregator-foundation-team.pcf.company.com:443`. The 443 is appended automatically. Under this route, there is the reverse proxy of the platform registered. The reverse proxy is terminating the SSL/TLS connection. Based on the hostname specified, the container of Promregator is detected and the request is forwarded to the port 443 (??? I would have expected that CF would expect that Promregator was running on 8080, though) of that container.
* Promregator then proxies the requests to the configured scraping targets accordingly (Single Target Scraping).

In short: I don't really understand how this works now, but leaving off "server.port: 443" would make sense to me.

**Comment by eaglerainbow** [comment]
> Thanks eaglerainbow - private message me if you'd like to do a one on one conversation and perhaps share a screen or two over teams. I'll meet at a time convenient for you. Also, I'm sure a few of my teammates would like to meet you.

We can do that independently. I will drop you a private mail...
